src/v/model/ktp.h

```diff
@@ -0,0 +1,347 @@
+     * patternrs are). By explicitly using std::string_view::operator==, as
+     * here, we get a recognized memcmp and a generally faster comparison.
+     */
+    static bool tp_equals(
+      const model::topic_partition& l, const model::topic_partition& r) {
+        return l.partition == r.partition
+               && std::equal_to<std::string_view>{}(l.topic(), r.topic());
+    }
+};
+} // namespace model
+
+namespace std {
+template<>
+struct hash<model::ktp> {
+    /**
+     * @brief The hash specialization for ktp.
+     *
+     * Compatible with the hash for ntp: a ktp will return the same hash as an
+     * ntp with the same topic, partition and the "kafka" namespace.
+     */
+    size_t operator()(const model::ktp& ktp) const {
+        return detail::ntp_hash(
+          model::kafka_ns_view, ktp.get_topic()(), ktp.get_partition());
+    }
+};
+} // namespace std
+
+namespace model {
+
+/**
+ * @brief A ktp object that remembers its hash code.
+ *
+ * In some contexts, a ktp/ntp object will be repeatedly hashed, which
+ * is a relatively expensive operation. Instead, we can swap in this type
+ * which remembers its hash code. It is a subclass of ktp, and can safely
+ * be passed as a reference to a ktp or sliced down to a ktp for convenience.
+ *
+ * The hash code is calculated once at constructon and never changes, except via
+ * assignment, since the underlying topic and partition cannot be changed (but
+ * this invariant must be enforced by the ktp class as noted there). Among other
+ * things this means that assignment of a ktp_with_hash from a ktp object does
+ * not compile: for this to be safe we'd need to recalculate the hash after
+ * assignment.
+ */
+struct ktp_with_hash : public ktp {
+    /**
+     * @brief Default-construct a ktp_with_hash object.
+     *
+     * Equivalent to an object with empty topic name
+     * and partition id 0.
+     */
+    ktp_with_hash()
+      : ktp_with_hash{"", 0} {}
+
+    /**
+     * @brief Construct a ktp_with_hash from topic and partition id.
+     *
+     * The namespace is implitly "kafka" and the hash is calculated at
+     * construction and cached within the object.
+     */
+    ktp_with_hash(model::topic t, model::partition_id id)
+      : ktp{std::move(t), id} {
+        update_hash();
+    }
+
+    /**
+     * @brief Weakly-typed constructor for ktp_with_hash.
+     */
+    ktp_with_hash(ss::sstring t, int32_t id)
+      : ktp_with_hash{model::topic(std::move(t)), model::partition_id(id)} {}
+
+    ktp_with_hash(const ktp_with_hash& rhs) = default;
+
+    /**
+     * @brief Move-construct a new ktp_with_hash object.
+     *
+     * This needs a non-default implementation since we need to update
+     * the rhs' hash after it has been moved-from.
+     */
+    ktp_with_hash(ktp_with_hash&& rhs) noexcept
+      : ktp{std::move(rhs)}
+      , _hash_code{rhs._hash_code} {
+        rhs.update_hash();
```

Comment:
Right, this same conversation played out on slack without a solid conclusion. 

In general it's kind of ill-defined the state of a moved-from object, but I'm wary of breaking invariants like `if (a == b) hash(a) == hash(b)` which this would do, and neither ntp nor ktp do that (all three classes share the odd moved-from state). It would mean that if such an object were put in a hash map it could not later be found or deleted by key, among other oddities.

Maybe a good compromise is to leave the hash inconsistent but have a debug-mode assert that we do not use the hash in this state.